src/rule.rs

```diff
@@ -0,0 +1,214 @@
+        &self.transform_rules
+    }
+
+    pub fn implement_rules(&self) -> &[RuleRef] {
+        &self.implement_rules
+    }
+}
+
+pub struct Binding<'a> {
+    pattern: &'a Pattern,
+    plan: &'a GroupPlan,
+    group_trace_id: usize,
+    group_plan_index: Vec<u32>,
+}
+
+impl<'a> Binding<'a> {
+    pub fn new(pattern: &'a Pattern, plan: &'a GroupPlan) -> Self {
+        Binding {
+            pattern,
+            plan,
+            group_trace_id: 0,
+            group_plan_index: vec![0],
+        }
+    }
+
+    fn extract_group_plan(&mut self, pattern: &Pattern, group: &Group) -> Option<GroupPlanRef> {
+        if pattern.is_leaf_or_multi_leaf() {
+            if self.group_plan_index[self.group_trace_id] > 0 {
+                self.group_plan_index.remove(self.group_trace_id);
+                None
+            } else {
+                Some(group.logical_plans()[0].clone())
+            }
+        } else {
+            let id = self.group_plan_index[self.group_trace_id];
+            if id as usize >= group.logical_plans().len() {
+                self.group_plan_index.remove(self.group_trace_id);
+                None
+            } else {
+                Some(group.logical_plans()[id as usize].clone())
+            }
+        }
+    }
+
+    fn matches(&mut self, pattern: &Pattern, group_plan: &GroupPlan) -> Option<Plan> {
+        if !pattern.match_without_child(group_plan) {
+            return None;
+        }
+
+        let mut inputs = Vec::new();
+        let mut pattern_index = 0;
+        let mut group_plan_index = 0;
+
+        while pattern_index < pattern.children().len() && group_plan_index < group_plan.inputs().len() {
+            self.group_trace_id += 1;
+            self.group_plan_index.resize(self.group_trace_id + 1, 0);
+
+            let group = &*group_plan.inputs()[group_plan_index].borrow();
+            let child_pattern = pattern.child(pattern_index);
+
+            let current_plan = self.extract_group_plan(child_pattern, group)?;
+            let child_plan = self.matches(child_pattern, current_plan.borrow().deref())?;
+            inputs.push(child_plan);
+
+            if !(child_pattern.is_multi_leaf()
+                && (group_plan.inputs().len() - group_plan_index > pattern.children.len() - pattern_index))
+            {
+                pattern_index += 1;
+            }
+            group_plan_index += 1;
+        }
+
+        Some(Plan::new(group_plan.operator().clone(), inputs))
+    }
+
+    fn next(&mut self) -> Option<Plan> {
+        if self.pattern.children().is_empty() && self.group_plan_index[0] > 0 {
+            return None;
+        }
+
+        let plan = None;
+
+        loop {
+            self.group_trace_id = 0;
+            if let Some(last) = self.group_plan_index.last_mut() {
+                *last += 1;
+            }
+
+            let plan = self.matches(self.pattern, self.plan);
+            if plan.is_some() || self.group_plan_index.len() != 1 {
```

Comment:
Is it the same logic as starrocks？while (expression == null && this.groupExpressionIndex.size() != 1)

Reply:
The condition should be `plan.is_some() || self.group_plan_index.len() == 1`